compiler/semtypes.nim

```diff
@@ -826,7 +827,8 @@ proc semRecordNodeAux(c: PContext, n: PNode, check: var IntSet, pos: var int,
       n[^1] = semConstExpr(c, n[^1])
       if n[^2].kind != nkEmpty:
         typ = semTypeNode(c, n[^2], nil)
-        typ = fitNodeConsiderViewType(c, typ, n[^1], n[^1].info).typ
+        let def = semExprWithType(c, n[^1], {}, typ)
+        typ = fitNodeConsiderViewType(c, typ, def, def.info).typ
       else:
         typ = n[^1].typ
```

Comment:
Wouldn't it make sense to sem the type node first (if there is any), then sem the default value based on the type? So something like:

```nim
let expectedType = if n[^2].kind != nkEmpty: semTypeNode(c, n[^2], nil) else: nil
n[^1] = semConstExpr(c, n[^1], expectedType = expectedType)
if expectedType != nil:
  n[^1] = fitNode(c, expectedType, n[^1], n[^1])
typ = n[^1].typ

Reply:
Thanks! I will merge it into my fix.

Reply:
included in https://github.com/nim-lang/Nim/pull/20716